src/compile/legend/assemble.ts

```diff
@@ -20,29 +26,178 @@ function setLegendEncode(
   legend.encode[part].update[vgProp] = vgRef;
 }
 
+/**
+ * Determines the underlying field name for a given scale channel within a model hierarchy.
+ * @param model - The model to search for the field definition
+ * @param channel - The scale channel (e.g., 'color', 'size', 'shape') to find the field for
+ * @returns The field name if found; otherwise undefined
+ */
+function getFieldKeyForChannel(model: Model, channel: ScaleChannel): string | undefined {
```

Comment:
I'm trying to remember whether we really need this. The model does have the fields already and I think we may already populate the field in the model. Can you check that we really don't have it yet? And if so, this method maybe should be in the model.